### PR TITLE
separates out data vs code shreds at the type level

### PR DESCRIPTION
**Pull request by behzadnouri**
```diff
- Context: This code is in preparation of 
- https://github.com/solana-labs/solana/pull/25237
- which implements a new shred variant with merkle tree 
- hashes of the erasure batch
```

#### Problem
Working towards revising shred struct to embed versioning so that a new
variant can contain merkle tree hashes of the erasure batch. To ease out
migration the commit adds more type-safety by distinguishing data vs
code shreds at the type level.

Additionally having both data and coding headers in each shred is
redundant as only one is relevant for each shred. The revised shred type
in this commit will only have one type-specific header.
https://github.com/solana-labs/solana/blob/c785f1ffc/ledger/src/shred.rs#L198-L203

#### Summary of Changes
Separated out data vs code shreds at the type level

### CODE REVIEW

File: ledger/src/shred.rs [hunk 1]
```diff
@@ -1028,146 +795,6 @@ mod tests {
         );
     }
 
-    #[test]
-    fn test_sanitize_data_shred() {
```
Comment (behzadnouri) [inline ledger/src/shred.rs:1032]: moved the test to `shred/legacy.rs`

File: ledger/src/shred.rs [hunk 2]
```diff
@@ -1028,146 +795,6 @@ mod tests {
...
-                Err(Error::InvalidDataSize {
-                    size: 1229,
-                    payload: 1228,
-                })
-            );
-        }
-    }
-
-    #[test]
-    fn test_sanitize_coding_shred() {
```
Comment (behzadnouri) [inline ledger/src/shred.rs:1100]: moved the test to `shred/legacy.rs`

File: ledger/src/shred/traits.rs [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    // Like Shred::erasure_shard but returning a slice.
+    fn erasure_shard_as_slice(&self) -> Result<&[u8], Error>;
+
+    // Portion of the payload which is signed.
+    fn signed_payload(&self) -> &[u8];
+
+    // Possibly zero pads bytes stored in blockstore.
+    fn resize_stored_shred(shred: Vec<u8>) -> Result<Vec<u8>, Error>;
+
+    // Only for tests.
```
Comment (jbiseda) [inline ledger/src/shred/traits.rs:29]: can `#[cfg(test)]` be used here?

File: ledger/src/shred/traits.rs [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        let flags = self.data_header().flags;
+        flags.contains(ShredFlags::DATA_COMPLETE_SHRED)
+    }
+
+    fn reference_tick(&self) -> u8 {
+        let flags = self.data_header().flags;
+        (flags & ShredFlags::SHRED_TICK_REFERENCE_MASK).bits()
+    }
+
+    // Only for tests.
```
Comment (jbiseda) [inline ledger/src/shred/traits.rs:74]: `[cfg(test)]` here?

File: ledger/src/shred.rs [hunk 5]
```diff
@@ -628,100 +448,87 @@ impl Shred {
...
+        let buffer = packet.data.get(OFFSET_OF_SHRED_SLOT..)?;
+        limited_deserialize(buffer).ok()
     }
 
     pub(crate) fn reference_tick_from_data(data: &[u8]) -> u8 {
-        let flags = data[SIZE_OF_COMMON_SHRED_HEADER + SIZE_OF_DATA_SHRED_HEADER
-            - size_of::<u8>()
-            - size_of::<u16>()];
-        flags & ShredFlags::SHRED_TICK_REFERENCE_MASK.bits()
+        // TODO: should check the type bit as well!
```
Comment (jbiseda) [inline ledger/src/shred.rs:487]: address TODO?

**Review by jbiseda — COMMENTED** [review]
nice refactoring!

File: ledger/src/shred/traits.rs (hunk 3 above)
Comment (behzadnouri) [reply ledger/src/shred/traits.rs:29]: These two functions are invoked outside of this crate; e.g:
https://github.com/solana-labs/solana/blob/24122768d/core/src/window_service.rs#L818
https://github.com/solana-labs/solana/blob/24122768d/core/src/repair_weighted_traversal.rs#L282

If we make these `#[cfg(test)]` then the functions will not be available outside the ledger crate.
`#[cfg(test)]` only works for tests within the same crate.

File: ledger/src/shred/traits.rs (hunk 4 above)
Comment (behzadnouri) [reply ledger/src/shred/traits.rs:74]: similarly here, the function is invoked outside of the crate:
https://github.com/solana-labs/solana/blob/24122768d/core/src/repair_weighted_traversal.rs#L283

File: ledger/src/shred.rs (hunk 5 above)
Comment (behzadnouri) [reply ledger/src/shred.rs]: done

File: ledger/src/shred/traits.rs (hunk 3 above)
Comment (jbiseda) [reply ledger/src/shred/traits.rs:29]: ah, right. I know this predates this PR, but would it be useful to prefix these functions wtih `testing_` or something to indicate the intent? feel free to disregard if this is going to cause churn between the refactoring PRs.

**Review by jbiseda — APPROVED** [review]

File: ledger/src/shred/traits.rs (hunk 3 above)
Comment (behzadnouri) [reply ledger/src/shred/traits.rs:29]: yeah, probably at some point once current changes are done, it would be good to rename these methods.

**Pull request merged**